nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/CuratorHeartbeatMonitor.java

```diff
@@ -0,0 +1,376 @@
+                removeHeartbeat(nodeId);
+                return;
+            }
+
+            // connection complete
+            clusterCoordinator.finishNodeConnection(nodeId);
+            clusterCoordinator.reportEvent(nodeId, Severity.INFO, "Received first heartbeat from connecting node. Node connected.");
+        }
+
+        if (heartbeat.isPrimary()) {
+            clusterCoordinator.setPrimaryNode(nodeId);
+        }
+    }
+
+
+    /**
+     * Fetches the latest heartbeats for each node from ZooKeeper.
+     * Visible for testing
+     */
+    Map<NodeIdentifier, NodeHeartbeat> fetchHeartbeats() {
+        logger.debug("Fetching heartbeats from ZooKeeper");
+        final List<String> nodeIds;
+        try {
+            nodeIds = curatorClient.getChildren().forPath(heartbeatPath);
+        } catch (final NoNodeException nne) {
+            logger.info("Could not find any heartbeats in ZooKeeper because the ZNode " + heartbeatPath + " does not exist");
+            return null;
+        } catch (final Exception e) {
+            logger.error("Failed to obtain heartbeats from ZooKeeper due to {}", e);
+            logger.error("", e);
+            clusterCoordinator.reportEvent(null, Severity.WARNING, "Failed to obtain heartbeats from ZooKeeper due to " + e);
+
+            return null;
+        }
+
+        logger.debug("Found {} nodes that have emitted heartbeats to ZooKeeper", nodeIds.size());
+        final Map<NodeIdentifier, NodeHeartbeat> latestHeartbeats = new HashMap<>(nodeIds.size());
+        for (final String nodeId : nodeIds) {
+            final HeartbeatMessage heartbeatMsg;
+
+            final String nodeHeartbeatPath = heartbeatPath + "/" + nodeId;
+            final Stat stat = new Stat();
+            try {
+                final byte[] serializedHeartbeatMsg = getClient().getData().storingStatIn(stat).forPath(nodeHeartbeatPath);
+                heartbeatMsg = (HeartbeatMessage) unmarshaller.unmarshal(new ByteArrayInputStream(serializedHeartbeatMsg));
+            } catch (final Exception e) {
+                logger.error("Failed to obtain heartbeat from ZooKeeper for Node with ID {} due to {}", nodeId, e);
+                logger.error("", e);
+                clusterCoordinator.reportEvent(null, Severity.WARNING, "Failed to obtain heartbeat for Node with ID " + nodeId + " from ZooKeeper due to " + e);
+
+                continue;
+            }
+
+            // update timestamp to be the timestamp that ZooKeeper reports
+            final long lastModifiedTime = stat.getMtime();
+            if (lastModifiedTime > latestHeartbeatTime) {
+                latestHeartbeatTime = lastModifiedTime;
+            }
+
+            latestHeartbeats.put(heartbeatMsg.getHeartbeat().getNodeIdentifier(), StandardNodeHeartbeat.fromHeartbeatMessage(heartbeatMsg, lastModifiedTime));
+            logger.debug("Received heartbeat from Node {}", nodeId);
```

Comment:
Do you think the above code (lines 341-348) could be moved inside of try block (after line 332) so you can remove continue? It may just be me, but it would be much easier to follow.